### PR TITLE
[NETBEANS-3335] Enhance robustness of HTML Lexer

### PR DESCRIPTION
**Pull request by matthiasblaesing**
After the update of the HTML lexer to support HTML5 style attributes,
new states in the lexer can be reached. Instead of bailing out with a
NullPointerException in HtmlLexer#equals, report comparisons with NULL
as false.

### CODE REVIEW

**Comment by neilcsmith-net** [comment]
Given comments on this, possible 11.2 patch update?

**Comment by geertjanw** [comment]
Yes, that would be good I think.

**Comment by matthiasblaesing** [comment]
I think this is safe for a point release. In the worst case the exception is exchanged for wrong syntax highlighting, but the user should then still be able to work with his/her file.

File: ide/html.lexer/src/org/netbeans/lib/html/lexer/HtmlLexer.java [hunk 1]
```diff
@@ -1552,8 +1552,12 @@ public void release() {
 
     /** @param optimized - first sequence is lowercase, one call to Character.toLowerCase() */
     private static boolean equals(CharSequence text1, CharSequence text2, boolean ignoreCase, boolean optimized) {
-        assert text1 != null : "text1 arg is null";
-        assert text2 != null : "text2 arg is null";
+        if(text1 == text2) {
```
Comment (junichi11) [inline ide/html.lexer/src/org/netbeans/lib/html/lexer/HtmlLexer.java]: Nitpick: should add whitespace behind `if`

File: ide/html.lexer/src/org/netbeans/lib/html/lexer/HtmlLexer.java [hunk 2]
```diff
@@ -1552,8 +1552,12 @@ public void release() {
...
+            return true;
+        }
+        if(text1 == null || text2 == null) {
```
Comment (junichi11) [inline ide/html.lexer/src/org/netbeans/lib/html/lexer/HtmlLexer.java]: Nitpick: should add whitespace behind `if`

**Comment by junichi11** [comment]
@matthiasblaesing 
Maybe, we can add unit tests to the smarty module:
```diff
diff --git a/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_01.tpl b/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_01.tpl
new file mode 100644
index 0000000000..c476b594d0
--- /dev/null
+++ b/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_01.tpl
@@ -0,0 +1 @@
+<a {if $smarty.server.REQUEST_URI|strstr:"/new"}class="underline"{/if} href="{url('new')}">
diff --git a/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_01.tpl.errors b/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_01.tpl.errors
new file mode 100644
index 0000000000..607da0e3dc
--- /dev/null
+++ b/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_01.tpl.errors
@@ -0,0 +1 @@
+Detected parser errors in the file:
diff --git a/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_02.tpl b/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_02.tpl
new file mode 100644
index 0000000000..f6bb7c35a4
--- /dev/null
+++ b/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_02.tpl
@@ -0,0 +1 @@
+<div class="{if $test eq ""}toto{/if}" {if $test}id="titi"{/if}></div>
diff --git a/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_02.tpl.errors b/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_02.tpl.errors
new file mode 100644
index 0000000000..607da0e3dc
--- /dev/null
+++ b/php/php.smarty/test/unit/data/testfiles/parserErrors/testParserNETBEANS3335_02.tpl.errors
@@ -0,0 +1 @@
+Detected parser errors in the file:
diff --git a/php/php.smarty/test/unit/src/org/netbeans/modules/php/smarty/editor/parser/TplParserErrorsTest.java b/php/php.smarty/test/unit/src/org/netbeans/modules/php/smarty/editor/parser/TplParserErrorsTest.java
index 1ef5f93591..1caa6f3434 100644
--- a/php/php.smarty/test/unit/src/org/netbeans/modules/php/smarty/editor/parser/TplParserErrorsTest.java
+++ b/php/php.smarty/test/unit/src/org/netbeans/modules/php/smarty/editor/parser/TplParserErrorsTest.java
@@ -118,6 +118,15 @@ public class TplParserErrorsTest extends TplTestBase {
         assertParserErrors();
     }
 
+    // no errors
+    public void testParserNETBEANS3335_01() throws Exception {
+        assertParserErrors();
+    }
+
+    public void testParserNETBEANS3335_02() throws Exception {
+        assertParserErrors();
+    }
+
     private String getTestFileRelPath() {
         return "testfiles/parserErrors/" + getName() + ".tpl";
     }

```

**Comment by matthiasblaesing** [comment]
@junichi11 thank you very much - the whitespace is fixed and your unittest is of course included (much appreciated).

**Comment by neilcsmith-net** [comment]
@matthiasblaesing assuming we are pushing some updates to 11.2, could you do a cherry picked PR on top of `release112` for this?  It will likely need a release-branch spec version increment in addition to the commits here (the point bit I guess, needs to still be less than master).  Be good if you can test building an NBM and checking it applies correctly on the release prior to opening.  Thanks!

**Pull request merged**